linkerd/stack/src/gate.rs

```diff
@@ -145,117 +163,210 @@ where
-        // If the gate is shut, wait for it to open by storing a future that
-        // will complete when it opens. This ensures that waiters are notified.
-        while self.rx.is_shut() {
-            trace!(gate.open = false);
-            if !self.is_waiting {
-                let rx = self.rx.clone();
-                self.waiting.set(async move { rx.opened().await });
-                self.is_waiting = true;
+        let permit = ready!(self.poll_acquire(cx));
+        ready!(self.inner.poll_ready(cx))?;
+        self.permit = Poll::Ready(Permit(permit));
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn call(&mut self, req: Req) -> Self::Future {
+        self.permit = match std::mem::replace(&mut self.permit, Poll::Pending) {
+            Poll::Pending => panic!("poll_ready must be called first"),
+            Poll::Ready(..) => Poll::Pending,
+        };
+
+        self.inner.call(req)
+    }
+}
+
+impl<S> Gate<S> {
+    fn poll_acquire<Req>(&mut self, cx: &mut Context<'_>) -> Poll<Option<OwnedSemaphorePermit>>
+    where
+        S: Service<Req>,
+    {
+        if !self.acquiring {
+            match self.rx.state() {
+                State::Open => return Poll::Ready(None),
+                State::Shut => {}
+                State::Limited(sem) => match sem.try_acquire_owned() {
+                    Ok(permit) => return Poll::Ready(Some(permit)),
+                    Err(TryAcquireError::NoPermits) => {}
+                    Err(TryAcquireError::Closed) => {
+                        tracing::warn!("Gate closed");
+                        return Poll::Pending;
+                    }
+                },
             }
-            ready!(self.waiting.poll_unpin(cx));
+
+            self.acquiring = true;
+            let mut rx = self.rx.clone();
+            self.acquire.set(async move { rx.acquire().await });
         }
 
-        debug_assert!(self.rx.is_open());
-        self.is_waiting = false;
-        trace!(gate.open = true);
+        let permit = ready!(self.acquire.poll_unpin(cx));
+        self.acquiring = false;
+        Poll::Ready(permit)
+    }
+}
 
-        // When the gate is open, poll the inner service.
-        self.inner.poll_ready(cx)
+impl<S> Drop for Gate<S> {
+    fn drop(&mut self) {
+        if let Poll::Ready(ref mut p) = self.permit {
+            // If the service is dropped while ready, return the permit to the
+            // semaphore.
+            p.relinquish()
+        }
     }
+}
 
-    #[inline]
-    fn call(&mut self, req: Req) -> Self::Future {
-        self.inner.call(req)
+// === impl Permit ===
+
+impl Permit {
+    fn relinquish(&mut self) {
+        drop(self.0.take());
+    }
+}
+
+impl Drop for Permit {
+    fn drop(&mut self) {
+        // Permits are forgotten so the `Tx` controller can decide when to allow
+        // more requests.
+        if let Some(p) = self.0.take() {
+            p.forget();
+        }
```

Comment:
nit, TIOLI: this seems needlessly complex --- instead of having the default `Drop` behavior of a `Permit` be to forget the inner semaphore permit, and having a `Drop` impl for `Gate` that releases the permit if it's held, can't we just remove all the `Drop` impls and `Permit::relinquish`, keep the existing `Drop` behavior, and have `Gate::call` explicitly call `forget` on the inner semaphore permit? that would, AFAICT, accomplish the same semantics, but much simpler.